### PR TITLE
Column coding strategies for Codable records

### LINKED ISSUE #1007: How to auto binding struct variable and table column name
### What did you expect to happen?

My DBA used "snake case" in create the data table (e.g., group_id),

but we are used "camel case" in design the struct (e.g., groupId),

is there a scheme that can be auto binding two way?

I know can be use "encode(to:)" to binding column,

but there will be a lot of duplicate code in each model,

could you help me solution this problem?

### PR DESCRIPTION
**Pull request by sberrevoets**
This adds an option to convert column names from snake case to camel
case, similar to how JSONDecoder works. In fact, the implementation here
is largely based on JSONDecoder's implementation.

- [x] This pull request is submitted against the `development` branch.
- [x] Inline documentation has been updated.
- [x] README.md or another dedicated guide has been updated.
- [x] Changes are tested.

Fixes #1007.

### CODE REVIEW

File: GRDB/Record/FetchableRecord.swift [hunk 1]
```diff
@@ -592,3 +610,80 @@ public enum DatabaseDateDecodingStrategy {
...
+    func convertedKey(key: String) -> String {
+        switch self {
+        case .useDefaultKeys:
+            return key
+        case .convertFromSnakeCase:
+            return Self.convertToSnakeCase(key)
+        }
+    }
+
+    fileprivate static func convertToSnakeCase(_ stringKey: String) -> String {
```
Comment (sberrevoets) [inline GRDB/Record/FetchableRecord.swift:641]: This implementation is copied verbatim from https://github.com/apple/swift-corelibs-foundation/blob/eec4b26deee34edb7664ddd9c1222492a399d122/Sources/Foundation/JSONEncoder.swift#L127-L175

Comment (groue) [reply GRDB/Record/FetchableRecord.swift:641]: Good idea 👍 

We should copy the [license](https://github.com/apple/swift-corelibs-foundation/blob/eec4b26deee34edb7664ddd9c1222492a399d122/Sources/Foundation/JSONEncoder.swift#L5-L9) somewhere, maybe inside the function body.

**Review by groue — COMMENTED** [review]
Thank you @sberrevoets, that's exactly how this had to be addressed 💯 

Don't you think encoding should be addressed as well? It's better when features are as complete as possible, and leave nothing obvious to be desired.

**Comment by sberrevoets** [comment]
I ran out of time for now but I'll try to spend some cycles on getting encoding to work as well.

**Comment by groue** [comment]
Thank you @sberrevoets. I'm looking forward for encoding support.

**Comment by Chen-Michael** [comment]
This is really good news.

**Comment by sberrevoets** [comment]
See the second commit for encoding support. I also modified the SwiftLint config a bit to not have to disable the linter for all long URLs, this felt a bit nicer.

Let me know how this looks, happy to squash the commits prior to merging!

File: .swiftlint.yml [hunk 2]
```diff
@@ -61,3 +61,6 @@ opt_in_rules:
 
 vertical_whitespace:
   max_empty_lines: 2
+
+line_length:
+  ignores_urls: true
```
Comment (groue) [inline .swiftlint.yml:66]: 🤩

File: GRDB/Record/FetchableRecord.swift [hunk 3]
```diff
@@ -614,3 +632,87 @@ public enum DatabaseDateDecodingStrategy {
...
+
+                // Next word starts at the capital before the lowercase we just found
+                wordStart = beforeLowerIndex
+            }
+            searchRange = lowerCaseRange.upperBound..<searchRange.upperBound
+        }
+        words.append(wordStart..<searchRange.upperBound)
+        let result = words.map({ (range) in
+            return stringKey[range].lowercased()
+        }).joined(separator: "_")
```
Comment (groue) [inline GRDB/Record/FetchableRecord.swift:715]: Not what I would expect for `convertFromSnakeCase` 😉 Do tests catch this?

**Comment by groue** [comment]
> See the second commit for encoding support.

Thank you very much 👍 I think the decoding code is not quite what it should be (see comment above about `joined(separator: "_")`)

> I also modified the SwiftLint config a bit to not have to disable the linter for all long URLs, this felt a bit nicer.

Thank you very much for this. I did not know this very useful SwiftLint option.

> Let me know how this looks, happy to squash the commits prior to merging!

It looks good, except the conversion from snake case which converts *to* snake case. This makes me wonder about tests as well. Please tell me if you need help.

File: GRDB/Record/FetchableRecord.swift (hunk 3 above)
Comment (sberrevoets) [reply GRDB/Record/FetchableRecord.swift:715]: It's mostly a naming issue. Let's say we're _decoding_ this `FetchableRecord`:

```swift
struct Record: FetchableRecord {
	static let databaseKeyDecodingStrategy = DatabaseKeyDecodingStrategy.convertFromSnakeCase
	var recordID: String
}
```

During the decoding process, the decoder gets the key as `recordID`, which we first need to convert _to_ snake case to look it up in the `Row` (which uses the SQL conventional `record_id`). It seems that `convertToSnakeCase` would be a better name then, but from a public API perspective that's not true:

```
struct Record: FetchableRecord {
	static let databaseKeyDecodingStrategy = DatabaseKeyDecodingStrategy.convertToSnakeCase
	var recordID: String
}
```

Now it reads as if `Record` will have its keys converted _to_ snake case, which is confusing as `recordID` is using camel case.

My suggestion is to make the public API as simple as possible, and possibly change some names around internally, but either way it's not super intuitive. Let me know what you prefer and I can make that change!

**Comment by groue** [comment]
@sberrevoets, there was no reply to my previous comment. You have no obligation to help of course, but I'd appreciate to know what are your plans, so that this PR does not remain stuck forever.

**Comment by sberrevoets** [comment]
I left a comment inline :)

**Comment by groue** [comment]
Thank you @sberrevoets, but the `convertFromSnakeCase` method still returns a string joined with `_`, when it should do the opposite. And this little mistake should have been caught by tests.

**Comment by sberrevoets** [comment]
That's what I was trying to explain above but I guess wasn't clear about. I don't fully understand the encoding/decoding processes so correct me if I'm wrong, but it seems like both processes start with the Record's fields as input keys. Those will always be camelcase, so both encoding and decoding processes need to convert them _to_ snake case.

This isn't super intuitive so it's a bit hard to explain, but if you run the tests I added in this PR I think you'll see what I mean—basically they aren't opposite/reverse processes of each other. See my comment above for more details, as this also has some implications on the naming of these public APIs.

**Comment by groue** [comment]
I get it now! Sorry I was numb, and didn't place me in your shoes enough. We sure have to convert from coding keys to columns in both coding and decoding, and we need the same conversion. 😅

Shouldn't we share the implementation and avoid code duplication?

**Comment by sberrevoets** [comment]
Possibly, yeah. It's not entirely clear to me how though, since you probably don't want to couple the encoding/decoding processes to each other. If you have a preference I'd be happy to make that change though.

**Comment by groue** [comment]
> Possibly, yeah. It's not entirely clear to me how though

With a shared function. All right @sberrevoets, you've done all the hard work, and this is great. I can finish this pull request.

**Comment by groue** [comment]
Hello @sberrevoets, will you please let me push to your branch, so that I can finish this PR?

File: GRDB/Record/EncodableRecord+Encodable.swift [hunk 4]
```diff
@@ -149,9 +150,9 @@ private class RecordEncoder<Record: EncodableRecord>: Encoder {
             } catch is JSONRequiredError {
                 // Encode to JSON
                 try autoreleasepool {
-
+
```
Comment (carltongibson) [inline GRDB/Record/EncodableRecord+Encodable.swift:153]: This looks like a missing editor setting to trim trailing white space?

Comment (groue) [reply GRDB/Record/EncodableRecord+Encodable.swift:153]: GRDB does not trim trailing white spaces.

Comment (carltongibson) [reply GRDB/Record/EncodableRecord+Encodable.swift:151]: OK, cool. Just makes the diff bigger is all. Your project, no problem. 👍

**Comment by carltongibson** [comment]
I just want to say, this is a great addition. Thanks for your work both on this @sberrevoets and @groue! 🎁

**Comment by groue** [comment]
The documentation of [JSONDecoder.KeyDecodingStrategy.custom](https://developer.apple.com/documentation/foundation/jsondecoder/keydecodingstrategy/custom) makes it clear that it turns a JSON key into a CodingKey.

Our current `DatabaseColumnDecodingStrategy` does the opposite: it turns a CodingKey into a column.

This is a smell. But not only: this makes our implementation of `allKeys` incorrect (https://github.com/groue/GRDB.swift/pull/1029/commits/86e4adf860aa029f9eb6a663e993d2b2fa430b90).

This is a blocker.

**Comment by groue** [comment]
All right, there is no real choice anyway.

Note that `DatabaseColumnDecodingStrategy.convertFromSnakeCase` will decode the `team_id` column into the `teamId`key, not the `teamID` key. This behavior is identical to `JSONDecoder.KeyDecodingStrategy.convertFromSnakeCase`:

```swift
import Foundation

let json = #"{ "id": 1, "team_id": 2 }"#
let decoder = JSONDecoder()
decoder.keyDecodingStrategy = .convertFromSnakeCase

do {
    // Success
    struct Player: Codable {
        var id: Int
        var teamId: Int
    }
    let player = try decoder.decode(Player.self, from: json.data(using: .utf8)!)
    
    // Player(id: 1, teamId: 2)
    print(player)
}

do {
    // Failure
    struct Player: Codable {
        var id: Int
        var teamID: Int
    }
    _ = try decoder.decode(Player.self, from: json.data(using: .utf8)!)
} catch {
    // No value associated with key CodingKeys(stringValue: "teamID",
    // intValue: nil) ("teamID"), with divergent representation teamId,
    // converted to team_id.
    print(error)
}
```

**Comment by groue** [comment]
All right, @sberrevoets, I now feel reasonably confident that we can merge this PR.

I can't really give advice for your future contributions, but you can try at looking at what I had to do until the PR was mergeable. One key idea, though, is:

After we (rightfully) agree that Foundation JSONDecoder and encoder strategies are the correct inspiration, we have to **do it fully**:

1. Foundation strategies come with a `custom` case. It is not quite okay to ignore it, and bring support for snake_case only. Users usually find in GRDB the features they expect, because this library aims at providing maximum coverage of use cases. It's never been about fulfilling the needs of a few privileged individuals only.

2. Foundation strategies come with an api design, and we can assume that Apple engineers are better at designing apis than we are. That's why we copy. And when we copy, let's do it fully, shamelessly, with our full heart! This is why I had to reverse the decoding strategy. And this is how users who already know custom JSON key decoding will feel at ease with custom database column decoding: this is *the same design*.

3. Foundation strategies come with two conversion methods (from and to snake_case). It has always been a smell that we needed only one. Only step 2 above could make us understand where the mistake was. This explains a lot why we talked past one each other, above. I just could not spot why copying Foundation strategies would not involve copying the two conversion methods.

I'm looking forward your next contribution. As you see, I'm always there to help good ideas thrive 👍

**Comment by groue** [comment]
Thank you @sberrevoets, column coding strategies have shipped in [v5.11.0](https://github.com/groue/GRDB.swift/tree/v5.11.0) 👍

**Comment by sberrevoets** [comment]
Awesome, thanks @groue!

**Pull request merged**